source/main.coffee

```diff
@@ -7,149 +7,53 @@ import * as util from "./util.coffee"
 { SourceMap } = util
 export { parse, generate, util }
 
-# Rules that are not cacheable
-# Essentially anything that depends on mutable state in the parser like indents and the rules that depend on them
-# One day this will be better supported by Hera
+import StateCache from "./state-cache.mts"
+
+# Need to no-cache any rule that directly modifies parser state
+# indentation stack, jsx stack, etc.
+
 uncacheable = new Set [
-  "ActualAssignment"
+  # Meta
+  "Debugger"
+  "Init"
+  "Program"
+  "Reset"
+
+  # Indentation
+  "InsertIndent"
+  "PushIndent"
```

Comment:
Shouldn't need this one either, because it calls `TrackIndented` that does the actual work. But I'm fine with keeping it.

Reply:
We do need both `PushIndent` and `TrackIndented`. If we cache `PushIndent` then sometimes we would use the cached one and not actually push the new indent after backtracking.

We need to cache the state modifying rules up to the point where they balance within a parent rule so we need to cache all the way up to `PushIndent` and `PopIndent` since they must be paired in the grammar.